### PR TITLE
[WIP] - module front

### LINKED ISSUE #293: 🚀 [REQUEST] - New mega menu module
### Summary

Introduction to new mega menu module.

Features: 
- [x] - BO - multiple elements types banner, custom link, category link, cms link, html content, product
- [x] - BO - possibility to enable/disable menu element
- [x] - BO - configuration to display menu item in mobile/desktop view
- [x] - BO - possibility to use custom css classes, 
- [ ] - BO - ~~allow main elements to be displayed as side menu elements~~ - version 1.1.0
- [x] - BO - build whole menu tree based on category item
- [x] - BO - add breadcrumb to navigate menu tree
- [ ] - BO - ~~add possibility to move one element to become other element children (restriction to prouduct, html content, banner)~~ - version 1.1.0
- [x] - FO - build cache key (id_shop, id_lang, id_group, random_key - that changes upon menu changes) and cache dynamic loaded menu elements with http get
- [x] - FO - desktop menu submenu elements fetched dynamically 
- [x] - FO - whole mobile menu fetched dynamically
- [x] - FO - new mobile/desktop menu UI
- [x] - BO - menu element desktop Bootstrap grid class selector

### Feature target required module/theme

is_megamenu

### Drawbacks

N/A

### PR DESCRIPTION
**Pull request by Oksydan**
This PR should close https://github.com/Oksydan/falcon/issues/293:

- FO - build cache key (id_shop, id_lang, id_group, random_key - that changes upon menu changes) and cache dynamic loaded menu elements with http get
- FO - desktop menu submenu elements fetched dynamically
- FO - whole mobile menu fetched dynamically
- FO - new mobile/desktop menu UI

### CODE REVIEW

File: _theme_dev/src/css/theme/components/_mobile-menu.scss [hunk 1]
```diff
@@ -0,0 +1,66 @@
+
+.mobile-menu {
```
Comment (kpodemski) [inline _theme_dev/src/css/theme/components/_mobile-menu.scss]: I'd consider something less generic there

File: is_mainmenu.php [hunk 2]
```diff
@@ -32,11 +33,17 @@ public function __construct()
         $this->bootstrap = true;
         parent::__construct();
 
-        $this->displayName = 'Main menu module';
-        $this->description = 'Main menu module';
+        $this->displayName = $this->trans('Main menu module', [], TranslationDomains::TRANSLATION_DOMAIN_ADMIN);
+        $this->description = $this->trans('Main menu module', [], TranslationDomains::TRANSLATION_DOMAIN_ADMIN);
```
Comment (kpodemski) [inline is_mainmenu.php:37]: Does the translation domain loaded from a different file work when you export translation for a module? i.e. if you want to build a release

File: src/Cache/TemplateCache.php [hunk 3]
```diff
@@ -25,14 +26,40 @@ class TemplateCache
...
     ) {
         $this->module = $module;
         $this->context = $context;
         $this->hookModuleRepository = $hookModuleRepository;
+        $this->configuration = $configuration;
+    }
+
+    public function getContextBasedRequestCacheKey(): string
+    {
+        return md5($this->getRequestCacheKey() . '|' . $this->context->language->id . '|' . $this->context->shop->id . '|' . $this->context->customer->id_default_group);
```
Comment (kpodemski) [inline src/Cache/TemplateCache.php]: Food for thought:
- availability of categories to certain categories is something less used in the ecosystem, that's my POV, it'd be interesting to have cache key configurable 👍🏻

File: src/Cache/TemplateCache.php [hunk 4]
```diff
@@ -25,14 +26,40 @@ class TemplateCache
...
+    }
+
+    private function getRequestCacheKey(): int
```
Comment (kpodemski) [inline src/Cache/TemplateCache.php]: Shouldn't that be a string as in `getContextBasedRequestCacheKey`?

```suggestion
    private function getRequestCacheKey(): string
```

Comment (kpodemski) [reply src/Cache/TemplateCache.php]: now that I've written that, I'm not sure anymore what was your idea for that ;-)

File: src/Cache/TemplateCache.php [hunk 5]
```diff
@@ -25,14 +26,40 @@ class TemplateCache
...
+    }
+
+    public function setNewRequestCacheKey(): void
+    {
+        $current = $this->getRequestCacheKey() + 1;
+
+        $this->configuration->set(static::CACHE_KEY, $current);
     }
 
     public function clearTemplateCache(): void
```
Comment (kpodemski) [inline src/Cache/TemplateCache.php:53]: Consider this in this method:

```
$hookedHooks = $this->hookModuleRepository->getAllHookRegisteredToModule($this->module->id);
$uniqueHooks = array_unique(array_column($hookedHooks, 'name'));
```

File: _theme_dev/src/js/theme/components/useHoverIntent.js [hunk 6]
```diff
@@ -0,0 +1,235 @@
+/**
+ * Creates a hover intent tracker for a given element.
```
Comment (kpodemski) [inline _theme_dev/src/js/theme/components/useHoverIntent.js:2]: Ha, I remember using jQuery hover intent years ago. Did you decide to create your own version or is that base on something?

File: src/Cache/TemplateCache.php (hunk 4 above)
Comment (Oksydan) [reply src/Cache/TemplateCache.php]: @kpodemski it's actually cache key for get request for front end ajax request. It's still work in progress. This cache key will be incremented very time any changes to menu will occur.

File: _theme_dev/src/js/theme/components/useHoverIntent.js (hunk 6 above)
Comment (Oksydan) [reply _theme_dev/src/js/theme/components/useHoverIntent.js:2]: @kpodemski yea, that's actually rewritten jQuery hoverIntent to vanilla.js but still I am not sure it's final implementation. Whole front end I made only for POC to find the best way to load submenu.

File: is_mainmenu.php (hunk 2 above)
Comment (Oksydan) [reply is_mainmenu.php:37]: @kpodemski it have no idea tbh 🤔. I will test it when everything is done.

File: _theme_dev/src/css/theme/components/_mobile-menu.scss (hunk 1 above)
Comment (Oksydan) [reply _theme_dev/src/css/theme/components/_mobile-menu.scss]: Thanks for suggestion :). Changes applied.

**Comment by Oksydan** [comment]
@kpodemski feel free to look at changes :). Front end design isn't final. I am still waiting for projects from UX designer.

**Pull request merged**